Docs: Update Development Dependencies Install

vfonic opened the pull request:
I've been following instructions on how to setup everything for contributing to Rails and I found it difficult to have to scroll through all OSes for each of the services needed to be installed. Instead, I grouped everything by OS and wrote above that which gem requires which additional service. I think this will get developers much faster to the point where they can start contributing.

Some sections were also huge and, in the middle of the section, start talking about separate tasks (which could be nicely broken into separate section). I broke down those sections into multiple sections. For example: [2.3 Set up and Run the Tests](http://edgeguides.rubyonrails.org/development_dependencies_install.html#set-up-and-run-the-tests) section. "Install Bundler" part could be moved to a separate section.

Some other remarks:

1. Please check if I wrote everything correctly and if I missed anything, if the nomenclature is correct (engine/service/etc.)
2. What's the convention about prepending CLI commands with `$` or `#` (or `mysql>`, etc.)?
It's hard to c/p these commands. Especially in the case of creating mysql user:

```bash
$ mysql -uroot -p

mysql> CREATE USER 'rails'@'localhost';
mysql> GRANT ALL PRIVILEGES ON activerecord_unittest.* to 'rails'@'localhost';
mysql> GRANT ALL PRIVILEGES ON activerecord_unittest2.* to 'rails'@'localhost';
mysql> GRANT ALL PRIVILEGES ON inexistent_activerecord_unittest.* to 'rails'@'localhost';
```

3. Should I just mention that macOS users can use `Brewfile`? I don't know what are the uses of some of the packages mentioned in the `Brewfile`.
4. Are there any efforts for redesigning the Guides website? I watched a Keynote talk by Nick Sutterer [Ruby is Dead - RedDotRubyConf 2017](https://youtu.be/4q3sycRJ5bs?t=2024) in which he mentioned how so many open source projects have beautiful websites. Not to say Rails Guides website is ugly, but there are certain parts that could use a little facelifting. (Maybe this is not needed :) )

From git commit message:
* Break down "Set up and Run the Tests" section into smaller parts
* Group documentation instructions by OS, not by gem
 * Although breakdown by gem allows for easier install of everything for a
particular gem, breakdown by OS allows for easier install of all services.
 * There's still a list of all additional services required for each of the
gems above the instructions for each OS.
* Remove instructions on how to run tests
 * Instead, refer the reader to continue reading "Contributing to Ruby on
Rails", which continues a section about running tests
* Update Bundler version referenced in "Ruby on Rails Guides Guidelines"

vfonic committed:
Docs: Update Development Dependencies Install

* Break down "Set up and Run the Tests" section into smaller parts
* Group documentation instructions by OS, not by gem
  * Although breakdown by gem allows for easier install of everything for a particular gem, breakdown by OS allows for easier install of all services.
  * There's still a list of all additional services required for each of the gems above the instructions for each OS.
* Remove instructions on how to run tests
  * Instead, refer the reader to continue reading "Contributing to Ruby on Rails", which continues a section about running tests

vfonic committed:
Update Bundler version referenced

oniofchaos commented on guides/source/development_dependencies_install.md:
````diff
@@ -43,215 +43,115 @@ $ git clone https://github.com/rails/rails.git
-```bash
-$ brew install sqlite3
-```
+* Action Cable depends on Redis
````
Might be better to link to each gem's README? For example, Action Cable: https://github.com/rails/rails/tree/master/actioncable#configuration

oniofchaos commented on guides/source/ruby_on_rails_guides_guidelines.md:
```diff
@@ -106,9 +106,7 @@ The guides and the API should be coherent and consistent where appropriate. In p
-Before generating the guides, make sure that you have the latest version of
-Bundler installed on your system. As of this writing, you must install Bundler
-1.3.5 or later on your device.
+Before generating the guides, make sure that you have the latest version of Bundler installed on your system. As of this writing, you must install Bundler 1.16.1 or later on your device.
```
Might be somewhere you can link to for the version # of bundler too.

oniofchaos replied:
Yeah, either rubygems or the releases section on github are the two ideas I have https://github.com/bundler/bundler/releases

vfonic replied:
I added the link to rubygems. I like releases page as well. It was a tough choice. :)
I went with rubygems, as that's the version you're actually going to be installing with rubygem's `gem` CLI.

dillonwelch reviewed (commented):
Just a couple thoughts on how to make the document more dynamic.

vfonic commented:
Hi @oniofchaos, thanks for the review!

I just got around to check your comments.

I'll rebase now and resolve merge conflicts. (Hope your comments don't get lost)

vfonic committed:
Add back changes lost after the merge conflict resolution

vfonic commented:
Hi @rafaelfranca,

I hope you're the right person to "ping".

It would be great to have someone look into this PR as it's been around for 5 months now. I understand it takes time so thank you for looking into it!

vfonic committed:
Link to Bundler gem on rubygems

rafaelfranca commented on guides/source/development_dependencies_install.md:
````diff
@@ -43,215 +43,121 @@ $ git clone https://github.com/rails/rails.git
 
 ```bash
-$ sudo pacman -S memcached
+$ brew install sqlite3 mysql postgresql redis memcached yarn imagemagick ffmpeg mupdf-tools poppler
````
If you are inside the Rails repository you can run just `brew bundle` and it will take care of everything.

https://github.com/rails/rails/blob/master/Brewfile

In fact I prefer that to this list here that will get outdated soon.

vfonic replied:
Thanks for feedback!

I mentioned that in the PR description above. I've replaced this code snippet with `brew bundle`. Please have a look.

vfonic committed:
Replace macOS instruction to install all dependencies with `brew bundle`

robin850 commented on guides/source/development_dependencies_install.md:
````diff
@@ -43,215 +43,120 @@ $ git clone https://github.com/rails/rails.git
 ```bash
-$ sudo pacman -S mariadb libmariadbclient mariadb-clients
-$ sudo pacman -S postgresql postgresql-libs
+# pkg install sqlite3 mysql56-client mysql56-server postgresql94-client postgresql94-server memcached imagemagick ffmpeg mupdf
````
That's not really important but `mysql80-{client, server}` and `postgresql11-{client, server}` are available on Fresh Ports. Just saying, may be it's not really important to match the latest version here. 🤷‍♂️

vfonic replied:
I believe this is important. We're advocating using the latest bundler, we should be advocating using the latest supported tools.

I've made these changes and also updated the link to MySQL troubleshooting documentation.

vfonic commented on guides/source/development_dependencies_install.md:
````diff
@@ -43,215 +43,120 @@ $ git clone https://github.com/rails/rails.git
-```bash
-$ sudo pacman -S memcached
-```
+WARNING: If you're working with Active Record code, you _must_ ensure that the tests pass for at least MySQL, PostgreSQL, and SQLite3. Subtle differences between the various adapters have been behind the rejection of many patches that looked OK when tested only against single adapter.
````
👍 I've removed the other paragraph and rephrased this one. I left this one in, as, it seems it has been added for a reason ("...behind the rejection of many patches")

vfonic commented on guides/source/development_dependencies_install.md:
````diff
@@ -43,215 +43,120 @@ $ git clone https://github.com/rails/rails.git
 ```bash
-$ brew install mysql
-$ brew install postgresql
+$ sudo yum install libsqlite3x libsqlite3x-devel mysql-server mysql-devel postgresql-server postgresql-devel redis memcached imagemagick ffmpeg mupdf
````
Hmmm I tried to find online as correct as I could which package names (and package managers) should be used. I don't have any of these OSes, so I tried to make the least amount of changes to that part of the documentation.

I'll replace SQLite package names. Perhaps switching yum with dnf could be something for another PR? I'm not sure if it's just a simple `yum` => `dnf` change or not.

vfonic committed:
Update package names and simplify rails gems dependencies documentation

vfonic commented:
@robin850  thanks for the great feedback!

robin850 commented:
> I'm not sure if it's just a simple `yum` => `dnf` change or not.

Yes, it is actually. According to the [Fedora wiki](https://fedoraproject.org/wiki/DNF), it "roughly maintains CLI compatibility with yum" but given we are only using the `install` sub-command we are all good but may be this change is not that important.

vfonic commented:
Yeah, I can make the change, but I'd rather have this merged before this PR get dragged on forever. This is a step forward. Let's iterate in small steps. It's not WIP work-in-progress, breaking change. But it IS hard to resolve merge conflicts for any PR that gets in before this one does...

robin850 commented on guides/source/development_dependencies_install.md:
````diff
@@ -43,215 +43,113 @@ $ git clone https://github.com/rails/rails.git
-
-```bash
-$ cd railties
-$ TEST_DIR=generators bundle exec rake test
````
Since you've removed all the testing instructions because they are all explained in the "Contributing" guide, it looks like this information is not present elsewhere. Maybe you can add a "Running all tests inside a directory" section to the contributing guide near the ["Running a single test"](https://guides.rubyonrails.org/contributing_to_ruby_on_rails.html#running-a-single-test) section. What do you think ?

robin850 reviewed (commented):
Yeah, you're right, let's try to merge that rather sooner that later. There is just a last concern, then I guess we're all good !

vfonic committed:
Move all test related documentation to "Contributing to Ruby on Rails" guide

vfonic committed:
Use dnf instead of yum for Fedora / CentOS packages installation [ci skip]

vfonic commented:
@robin850 
I went over the whole changeset again:
* Rebased on top of `master` (hope there are no comments lost)
* Removed the two bullet points no longer relevant from dev dependencies install
* Improved "Contributing to Rails" documentation about running a specific directory and file (although file is already briefly mentioned below)
* Updated installation instructions for Fedora/CentOS to use dnf instead of yum (also went to yarn's official documentation website to check instructions)

vfonic commented:
Awesome! Thank you for taking care of this! 🤘 ❤️
